OgreMain/src/block_n.hpp

```diff
@@ -0,0 +1,219 @@
+    auto buffer = static_cast<unsigned char*> (dst) + block_bytes<t> (newsize);
+    block_relocate_memcpyable (newsize, moveable, buffer, args...);
+}
+
+template <typename t>
+void block_construct (unsigned size, t* t1)
+{
+    for (auto i = 0u; i < size; ++i)
+        new (t1 + i) t{};
+}
+
+template <typename t, typename... args_t>
+void block_construct (unsigned size, t* t1, args_t*... args)
+{
+    block_construct<t> (size, t1);
+    block_construct (size, args...);
+}
+
+template <typename t>
+void block_destruct (unsigned size, t* t1)
+{
+    for (auto i = 0u; i < size; ++i)
+    {
+        auto ptr = t1 + i;
+        ptr->~t ();
+    }
+}
+
+template <typename t, typename... args_t>
+void block_destruct (unsigned size, t* t1, args_t*... args)
+{
+    block_destruct<t> (size, t1);
+    block_destruct (size, args...);
+}
+
+template <typename t>
+void block_free (t* t1)
+{
+    auto ptr = static_cast<void*>(t1);
+    free (ptr);
+}
+
+template <typename t, typename... args_t>
+void block_delete (unsigned& size, t* t1, args_t*... args)
+{
+    if (!size) return;
+    block_destruct (size, t1, args...);
+    block_free (t1);
+    size = 0;
+}
+
+template <typename t>
+void block_delete_memcpyable (unsigned& size, t* t1)
+{
+    if (!size) return;
+    block_free (t1);
+    size = 0;
+}
+
+template <typename... args_t>
+void block_malloc (unsigned size, args_t**... args)
+{
+    auto bytes = block_bytes <args_t...> (size);
+    auto mem = malloc (bytes);
+    block_assign (size, mem, args...);
+}
+
+template <typename... args_t>
+void block_new (unsigned size, args_t**... args)
+{
+    if (!size) return;
+    block_malloc (size, args...);
+    block_construct (size, *args...);
+}
+
+template <typename owner_t, typename t, typename... args_t>
+void block_resize (owner_t& original, unsigned& size, unsigned newsize, t** t1, args_t**... args)
+{
+    if (newsize == 0)
+    {
+        block_delete (size, *t1, *args...);
+        return;
+    }
+    auto bytes = block_bytes <t, args_t...> (newsize);
+    auto mem = malloc (bytes);
+    auto moveable = std::min (size, newsize);
+    block_relocate (newsize, moveable, mem, t1, args...);
+    block_destruct (size, *t1, *args...);
+    block_free (*t1); // free the old allocation
+    size = newsize;
+    block_assign (size, mem, t1, args...);
+}
+
+template <typename owner_t, typename t, typename... args_t>
+void block_resize_memcpyable (owner_t& original, unsigned& size, unsigned newsize, t** t1, args_t**... args)
+{
+    if (newsize == 0)
+    {
+        block_delete_memcpyable (size, *t1);
+        return;
+    }
+    auto bytes = block_bytes <t, args_t...> (newsize);
+    auto mem = malloc (bytes);
+    auto moveable = std::min (size, newsize);
+    block_relocate_memcpyable (newsize, moveable, mem, t1, args...);
+    block_free (*t1); // free the old allocation
+    size = newsize;
+    block_assign (size, mem, t1, args...);
+}
+
+template <typename... args_t>
+struct block_n
```

Comment:
is this any different than `std::vector<std::tuple<args...>>`?

Reply:
i think the vector<tuple> approach is similar although i'm not sure of the exact syntax.  for one, you'd be referring to indexes (ex. 0, 1, 2) instead of variable names (ex. 'TTLs').  1. block_n uses a single allocation for all of the arrays and therefore all of the arrays are ajacent in memory.  i.e. the TTLs array is immediatly followed by the translations array which is followed by the rotations data... ex. <TTL><TTL><TTL><POS><POS><POS><ROT><ROT><ROT> etc.  so when particle "expire" loop is called, it decrements all the TTLS.  then each <POS> is updated when velocity is applied, followed by each of the <ROT> etc etc. 2. block_n does a memcpy instead of std::move() which for classes (ex. Vector3) ensures ctors() and dtors() are called which we don't care about in this case.  3. vector slightly larger footprint.  4. i use block_n for my ECS classes and the particle work is a rapid prototype so it was easier to use.  (note if you remove the "_memcpyable" then block_n supports move operations instead of memcpy).

Reply:
the debug drawing API refactoring is an *excellent* example of what needs to be done everywhere in ogre.
i..e. pulling out unrelated parts and keeping them separate as to avoid polluting the runtime code.  for example, the particle systems, emitters, affector runtime code should be separate from the script parsing, particle system manager templates as singletons, etc.

Reply:
(and i think this is my issue with unreal, unity, godot, ogre 2.x+ is that it feels tightly coupled.  it's beautiful and super fast, but i'd rather mix and match what i get, and have a loosely coupled system where i pick and choose from what i want)

Reply:
> we could move the entire legacy particles to the FX plugin.

that would be a "ParticleFX" component, as you typically do not use any Plugin specific API. The current PluginFX implementation provides affectors and emitters that are only exposed through the common StringInterface property API.
This is also why Particle is needed in Main - to define the common API to be used by Plugins.

> that would be FX3 plugin and the old FX and FX2 would just never get upgrades anymore.

the thing is, that nobody likes updating their **working code** - no matter what the benefit is. You said yourself, you would like to have the particle thing cleared up in order to never have to thing about it again.
Imagine having to port it to FX2, with a FX3 at the horizon.

However, I am quite confident that we can come up with an *public* API that will suit a FX3 implementation at lets say 90% of the optimal performance.
The key ingredient is probably just switching from list<Particle*> to vector<Particle>. With the latter iteration is fast and we can just memcpy it (which is cheap) to a hardware buffer or whatever will be needed down the road.

Reply:
i was imagining you would update your plugin.cfg to use the particles you want and set the #include headers to look in the correct folder.  i.e. just a compiler and runtime configuration change, not break existing code.

Reply:
>  set the #include headers to look in the correct folder.

if you #include the plugin headers, you have to actually link against the plugin - at which point it is no longer a plugin, but an ordinary library. Thats why I would rather call it "Component".

Reply:
ah gotcha. makes sense.